samples/interactive-tutorials/user_environment_setup.sh

```diff
@@ -14,61 +14,83 @@
+  # Log out of service account
+    gcloud auth revoke
+fi
+echo "$service_account_id"
+
+editor=$(gcloud projects get-iam-policy $project_id  \
+--flatten="bindings[].members" \
+--format='table(bindings.role)' \
+--filter="bindings.members:$service_account_id ROLE=roles/editor")
+
+retail_admin=$(gcloud projects get-iam-policy $project_id  \
+--flatten="bindings[].members" \
+--format='table(bindings.role)' \
+--filter="bindings.members:$service_account_id ROLE=roles/retail.admin")
+
+
+# assign necessary roles to your new service account
+# Check if any of the needed roles is missing
+if [ -z "$editor" ] || [ -z "$retail_admin" ]
+then
+    # Assign necessary roles to your new service account.
+    for role in {retail.admin,editor}
     do
-      gcloud projects add-iam-policy-binding "$project_id" --member="serviceAccount:$service_account_id@$project_id.iam.gserviceaccount.com" --role=roles/"${role}"
-  done
+        gcloud projects add-iam-policy-binding "$project_id" --member="serviceAccount:$service_account_id@$project_id.iam.gserviceaccount.com" --role=roles/"${role}"
+    done
+    echo "Wait ~60 seconds to be sure the appropriate roles have been assigned to your service account"
+    sleep 60
+fi
 
-} && {
+# upload your service account key file
+service_acc_email="$service_account_id@$project_id.iam.gserviceaccount.com"
+gcloud iam service-accounts keys create ~/key.json --iam-account "$service_acc_email"
 
-  echo "Wait ~60 seconds to be sure the appropriate roles have been assigned to your service account"
-  sleep 60
+# activate the service account using the key
+gcloud auth activate-service-account --key-file ~/key.json
 
-  # upload your service account key file
-  service_acc_email="$service_account_id@$project_id.iam.gserviceaccount.com"
-  gcloud iam service-accounts keys create ~/key.json --iam-account "$service_acc_email"
 
-  # activate the service account using the key
-  gcloud auth activate-service-account --key-file ~/key.json
+# install needed Google client libraries
+cd ~/cloudshell_open/java-retail/samples/interactive-tutorials || exit
+mvn clean install -DskipTests
 
-} && {
 
-  # install needed Google client libraries
-  cd ~/cloudshell_open/java-retail/samples/interactive-tutorials || exit
-  mvn clean install -DskipTests
-
-} && {
-
-  # Print success message
-  echo "========================================"
-  echo "The Google Cloud setup is completed."
-  echo "Please proceed with the Tutorial steps"
-  echo "========================================"
-
-} || {
-
-  # Print error message
-  echo "========================================"
-  echo "The Google Cloud setup was not completed."
-  echo "Please fix the errors above!"
-  echo "========================================"
-}
+# Print success message
+echo "========================================"
+echo "The Google Cloud setup is completed."
+echo "Please proceed with the Tutorial steps"
+echo "========================================"
```

Comment:
Does this "The Google Cloud setup is completed." message appear regardless when any command above fails (along with "The Google Cloud setup was not completed.")?

Reply:
No, this message only appears if everything is ok, otherwise error function stops the execution of the script and it never gets to this message.